### PR TITLE
BFD-2476: Execute Partially Adjudicated Smoketest During Deployments

### PR DESCRIPTION
**Pull request by malessi**
**JIRA Ticket:**
[BFD-2746](https://jira.cms.gov/browse/BFD-2746)

**User Story or Bug Summary:**

As a BFD Engineer I want the Regression Suite to exercise the BFD Server as fully as possible in order to test for regressions in both performance and behavior

---

### What Does This PR Do?

This PR adds the PACA smoke tests to the existing Regression Suite Lambda such that they will run immediately after the existing Regression Suite during deployment and so the outcome of the smoketest will influence the result of the corresponding deployment stage.

These smoke tests differ from the existing Regression Suite in that they use random, non-synthetic data and thus do not have consistent performance characteristics from environment-to-environment or run-to-run; thus, statistics from these smoke tests are _not_ collected as part of our existing Regression Suite performance baseleine. Only the final validation result is considered with respect to the PACA smoketests, and validation fails if there are _any_ failing test cases. These tests will _always_ run with the same static configuration of 10 synthetic users for 10 seconds.

It is expected that a more comprehensive set of PACA endpoint tests will be _merged_ with the existing Regression Suite when BFD-2490 is complete. This is currently blocked until a consistent set of synthetic data is available to test against, as _performance_ regression testing requires such data.

In addition to adding the smoke tests to the Regression Suite Lambda, several code improvements have been made:

1. The entrypoint for stats-related operations, in `bfd_user_base.py`, has been refactored such that the code's behavior is more obvious with respect to failure cases and conditional logic
2. The cases where the Locust process exits with error code `1` have been made much more obvious and have been centralized to `bfd_user_base.py`
3. Multiple type hinting errors and linting errors have been corrected
4. Pylance/Pyright's `strict` type-checking mode has been enabled with some rules explicitly disabled to compromise between safety and practicality (as several of the libraries used do not export proper types, or encourage "best practices" that trigger some rules)
5. Multiple errors with respect to the Regression Suite Lambda Docker image's build process have been corrected:
   1. The `Dockerfile.dockerignore` is now actually respected
   2. Authentication errors with Amazon's Public ECR have been corrected when using `build-push.sh`
   3. Unnecessary image tag pushes have been minimized

This PR has been tested by building a new image and applying it to `test`'s Regression Suite Lambda and verifying that, when run, the Lambda runs both the Regression Suite and PACA smoketests.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

- Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

- If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

  - Does this PR add any new software dependencies?
    - [ ] Yes
    - [x] No
  - Does this PR modify or invalidate any of our security controls?
    - [ ] Yes
    - [x] No
  - Does this PR store or transmit data that was not stored or transmitted before?
    - [ ] Yes
    - [x] No

- If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
  - Do you think this PR requires additional review of its security implications for other reasons?
    - [ ] Yes
    - [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

- N/A

### Submitter Checklist

I have gone through and verified that...:

- [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
- [x] This PR is reasonably limited in scope, to help ensure that:
  1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
  2. There isn't too much of a burden on reviewers.
  3. Any problems it causes have a small "blast radius".
  4. It'll be easier to rollback if that becomes necessary.
- [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
- [x] The data dictionary has been updated with any field mapping changes, if any were made.
- [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
- [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
- [x] Reviews are requested from both:
  - At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
  - Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
- [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
  - Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/utils/locust_tests/common/bfd_user_base.py [hunk 1]
```diff
@@ -161,7 +182,11 @@ def get_by_url(self, url: str, headers: Optional[Dict[str, str]] = None, name: s
             catch_response=True,  # type: ignore -- known Locust argument
         ) as response:
             if response.status_code != 200:
-                response.failure(f"Status Code: {response.status_code}")
+                if isinstance(response, ResponseContextManager):
+                    # pylint: disable=E1121
```
Comment (JeremyIR) [inline apps/utils/locust_tests/common/bfd_user_base.py:186]: Im not as familiar with pylint. What purpose does this comment serve?

Comment (malessi) [reply apps/utils/locust_tests/common/bfd_user_base.py:186]: At this point, `response` can be either an instance of `ResponseContextManager` or `FastResponse`. `ResponseContextManager` implements _both_ `failure()` and `failure(exception: str | Exception)` whereas `FastResponse` implements _only_ `failure()`. Therefore, prior to this change, _both_ Pyright/Pylance and Pylint were reporting a possible error here where if `response` was a `FastResponse` calling `response.failure("...")` would throw.

Unfortunately, only Pylance/Pyright seems to understand that line 185 implicitly _coerces_ `response` to `ResponseContextManager` whereas Pylint seems to still believe that `response` could be `FastResponse` if the `isinstance` check is `true`. So, to stop Pylint from reporting false positives, the rule that corresponds with that particular error ([`too-many-function-args`](https://pylint.pycqa.org/en/latest/user_guide/messages/error/too-many-function-args.html)) is disabled using the comment on line 186.

Hopefully that gives some context as to what the comment does.

Comment (JeremyIR) [reply apps/utils/locust_tests/common/bfd_user_base.py:186]: Got it

**Review by mjburling — DISMISSED** [review]
Logic appears sound and the container builds _just_ fine with some nice little improvements along the way. LGTM.

**Review by mackec1 — APPROVED** [review]
LGTM

**Review by JeremyIR — DISMISSED** [review]
LGTM

**Pull request merged**